crates/aptos-in-memory-cache/src/caches/sync_mutex.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{SizedCache, SizedCacheEntry};
+use parking_lot::Mutex;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+
+type CacheEntry<T> = Arc<Mutex<Option<SizedCacheEntry<T>>>>;
+
+const DEFAULT_MAX_NUM_CACHE_ITEMS: usize = 1_000_000;
+
+/// An in-memory cache that uses a mutex to synchronize access to the cache entries.
+/// The cache is designed with indexing use cases in mind and offers a deterministic FIFO eviction policy.
```

Comment:
What if we insert out of order, e.g. in the case of a backfill? We'll be kicking out the more recent items right?

Reply:
Why would backfilling be a concern? It's not like as we are adding live data to the cache, we will all of a sudden decide to add backfill data to it too...

The assumption we make is that the numerical deviation in the keys of out of order inserts is no where near the cache capacity so that out of order inserts will not evict new values after wrapping around.